### ISSUE TITLE
Add ability to mark messages unread

### ISSUE THREAD

**Issue by colans** [vector-im/element-meta#347]
As in other messaging applications, it would be nice to mark a particular message in a room as unread, returning to the state before it was read.  This is helpful for when you need a future reminder to do something, or you didn't have a chance to read everything up to the present, and would like to move the unread marker back.

So when clicking on the 3 dots to open the message menu, ideally there would be an item entitled *Mark unread*.

*Amendment (2021-11-30): I'd be fine with being able to mark the entire room/conversation unread.  Signal did this, and it works well enough for me so I believe it could work here as well, even though it would be slightly less than optimal.*

**Comment by Throne3d** [comment]
(Copied from <https://github.com/vector-im/riot-web/issues/8666#issuecomment-464410039>, where I put it first:)

From #riot-web:matrix.org:
> **Throne3d**, 02:35
> Discord uses alt+clicks to let you manually mark some point read/unread, so you can move your unread marker to save your place if you're reading history
> (it's not very well advertised as a feature)
> I find that very useful – maybe it'd be a good thing to add wrt this?

(If you alt+click the last message in a channel, it marks the whole channel as read – in fact, it marks everything up to, and including, the message you click, as read, and everything after it as unread.)

**Comment by jamesmontalvo3** [comment]
Just making sure we’re talking about the same thing: riot shouldn’t allow individual messages marked unread; it should allow you to select a message and mark it “unread” which just moves the unread pointer to that message, implying everything before that is read and everything after unread. I’m not expecting specific messages to be unread interspersed with read messages.

**Comment by aaronraimist** [comment]
@MightyCreak yes a public bounty can be added using websites like https://www.bountysource.com/. They had asked how to move it up in the roadmap. The roadmap is for the Element team so to move it up in the roadmap that’s who you need to talk to. 

@nunoperalta yes that’s true however Synapse is also open source. Anyone can contribute code to Element and anyone can contribute code to Synapse so anyone could implement this feature.

**Comment by yousefamar** [comment]
> The issue is we are swamped with paid work already compared to the engineering resources we have

Wouldn't the public bounty suggestion solve this problem? Either way, IMO, this is a good problem to have — you can now hire more developers. In fact, it sounds like you should be _preemptively_ hiring because this is only going to get worse. If the issue is the overhead in managing all that, I think these bounty programmes are exactly suited for that, as you can use them to incentivise _external_ developers to contribute, right? I don't think it's a good idea for non-maintainers to manage a bounty programme on your behalf, that's my 2 cents.

**Comment by d33tah** [comment]
I skimmed through the comments here and wanted to underline something: lack of presence for this feature can and **will affect the mental health of users**. I'm not exagerrating. When using Element, it's way more often than you imagine that you get a message sometime late during the day and have to decide whether read it and react immediately or skip it and regret because it was important. Given the pandemic, it's something that has even more importance. Especially since there's no way that I can see to turn off read notifications (see vector-im/element-web#2527), so it's a pretty nasty combination, creating unnecessary stress and miscommunication.

**Comment by gwpl** [comment]
The other way to to could be to be able to "bookmark" some messages and
they appear.then in "saved"/"bookmarked" (as.on slack)

**Comment by ShadowJonathan** [comment]
That's a whole separate feature; vector-im/element-meta#319

**Comment by SethFalco** [comment]
When or if this is implemented, it would be nice if we could hold `ALT` and click a message to mark all messages up-to that one selected as unread.

Both Discord and Slack have this behavior.

With both Discord and Slack, it works like a marker. For example, it doesn't matter where the last unread message was, when I hold `ALT` and click, it will mark anything above it as read, and anything below it as unread. (Even if messages above it were previously unread.)

**Comment by SethFalco** [comment]
One thing to also think about, how will this impact read receipts? Could we have an option to take back read receipts, or for the read receipt to have an effect on it if it was marked unread after?

In my opinion, it'd be great if read receipts turned gray if the message was opened then marked as unread.

This should probably be an option, not sure if it would be enabled or disabled by default.

The motivation is for cases where we want to come back to a message later, but in some cases, the other user may feel like they were "left on read".

**Comment by t3chguy** [comment]
Mark as read in Matrix is the read receipt, I haven't read the MSC but I'd presume it'd back pedal the RR. The RM on the other hand is the last point in history you've read everything up to.

**Comment by ShadowJonathan** [comment]
> but I'd presume it'd back pedal the RR.

@t3chguy [MSC2867](https://github.com/matrix-org/matrix-doc/pull/2867) doesn't do that, it just adds an extra bit of information to room account data to mark it being unread, it doesn't revert read receipts, which would be a much more difficult task.

**Comment by mcg-matrix** [comment]
> > but I'd presume it'd back pedal the RR.
> 
> [...], it doesn't revert read receipts, which would be a much more difficult task.

Just a thought: If the status of a message having been marked unread is stored for the account but does not affect anybody else's view of the world, it's just one special application of being able to mark (or flag or star or pick-a-verb) messages somehow, isn't it?

I think it would be great if I could add an indicator to funny messages, a different indicator (different colour, shape, sign, ...) to messages that need a follow-up sometime later, a third style indicator to messages that need a follow-up soon (= mark as unread), and so on.
 - For one (or multiple) selected indicator, show a count in the list of rooms and make it easy to find the marked messages in a room, and the application of “marking messages unread” should be covered.
 - Add a feature of filtering the list of rooms (list of threads in a room, list of messages in a room, ...) based on the indicators, and we're getting much closer to having a fine tool for asynchronous work. :-)

**Comment by MightyCreak** [comment]
> Just a thought: If the status of a message having been marked unread is stored for the account but does not affect anybody else's view of the world, it's just one special application of being able to mark (or flag or star or pick-a-verb) messages somehow, isn't it?

Yes, but you want your unread marker to propagate across your different sessions (desktop, web, mobile, ...), that's why it has to be stored server-side (i.e. the account) and not only client-side (i.e. the application).

**Comment by ChargingBulle** [comment]
I would like to use this feature as "not yet read"
because sometimes you accidentally "read" something.
For example when you open the Element Electron app and it opens the last visited channel and thus marks that as "read".

Slack (and others) handle this much better. When someone sends me contract work to review perhaps I only have time for it in the next week. Slack offers a remind-me-again feature which works really great for "Yes I read it but no I'll act upon it later"

**Comment by nailyk-fr** [comment]
Hello, 

I didn't read all but see that you are talking about a bounty. Is there one already in place ? 
If not, is it ok I start one ? 

Thanks for your answer and sorry for my 'noisy' comment.

**Comment by gabrc52** [comment]
Here's some random person's musings on the importance of this feature: https://matthewbischoff.com/mark-as-unread-for-messages/

**Comment by yousefamar** [comment]
This is one of the main reasons I use Schildichat

**Comment by nerdoc** [comment]
@yousefamar: Why? Schildichat does not provide this feature [yet](https://github.com/SchildiChat/schildichat-desktop/issues/109).

**Comment by yousefamar** [comment]
@nerdoc To clarify, I meant marking a _room_ as unread, not an individual message, which Schildichat supports on both desktop and mobile. On mobile you even have the option "open without reading". Frankly, I can't believe Element still doesn't have mark unread -- would have hoped the feature could be pulled from Schildichat since it's a fork afaik.

**Comment by nerdoc** [comment]
ah, OK, I see. yes,, Schildichat can mark rooms as unread, which is IMHO the feature I need. individual messages doesn't make sense IMHO, as I don't want to mark the third-last message as unread. It's always the last one, hence the room.

**Comment by gabrc52** [comment]
Yeah, tbh I just want to be able to mark conversations as unread just so that I can stop accidentally ignoring friends and classmates. Mark as unread is an indication that I should get back to it later.

Putting the burden on the user to remember to respond to conversations is just going to mean less conversations are going to be responded to, and overall less talking on the Matrix network.

**Comment by MightyCreak** [comment]
> ah, OK, I see. yes,, Schildichat can mark rooms as unread, which is IMHO the feature I need. individual messages doesn't make sense IMHO, as I don't want to mark the third-last message as unread. It's always the last one, hence the room.

I do have a use case for "mark message as unread". I use it a lot in other apps (e.g. EMail, Teams, ...). It's especially useful in rooms with several people.

Example: you enter a room where you have 15 unread messages. Most of the time, you can skim through, but not this time: there's a message that picked your interest. You decide to mark the message as unread (from this point in time, not just this message especially), so you can close the app and come back when you'll have more time to read it and its responses, and think about it.

I'd say I do that at least once per day in the other apps.

**Comment by gabrc52** [comment]
Yes, that use case does make sense for large rooms, and I do use it on Discord.

Although I think that the point is that having the feature is better than not at all. This issue has been open since 2017 and it appears to be blocked only because of this doubt of whether to mark rooms or messages as unread. It is now 2024.

**Comment by Jerome-Herbinet** [comment]
I don't understand at all how a feature as basic and essential as this isn't implemented in Element (whereas it is in Rocket.Chat). My colleagues and I who use Element on a daily basis find this very annoying.

**Comment by t3chguy** [comment]
Element Web & Desktop already have this.

<img width="398" alt="image" src="https://github.com/element-hq/element-meta/assets/2403652/cc49b915-8839-4a6f-921a-18b4bb554de1">

**Comment by nunoperalta** [comment]
Thank you. But I think it would be much more convenient to allow choosing a message (either right-click or long-tap on a message I choose) and "Mark Unread from Here". Skype has this functionality.

Just marking a whole thread as unread is not as useful as marking the exact position of that thread that I need to read from.

**Comment by t3chguy** [comment]
Sure, then I suggest you open an issue in the Matrix spec to support that. Right now the read receipts/markers can only move forwards, the server will block movement backwards

**Comment by nunoperalta** [comment]
My understanding is that this is the one for that:
https://github.com/matrix-org/synapse/issues/2632

**Comment by t3chguy** [comment]
No, that's just asking the server to provide a count of unread messages, nothing about allowing the markers to move backwards in time. https://github.com/matrix-org/matrix-spec is the repo you want

**Comment by nunoperalta** [comment]
Thank you. Submitted - https://github.com/matrix-org/matrix-spec/issues/1893

**Comment by colans** [comment]
> Element Web & Desktop already have this.

Wow, amazing!  When did that get in?  Anyway, I'm happy with closing this as soon as it gets in the mobile clients.

**Comment by RNLFoof** [comment]
That's a screenshot of marking an entire room as unread, while this issue was about specific messages, no?

**Comment by Jerome-Herbinet** [comment]
> That's a screenshot of marking an entire room as unread, while this issue was about specific messages, no?

@RNLFoof you are right ; marking the whole channel as unread is interesting, but not great (not enough from an UX point of view). Users need to mark specific messages as unread.

**Comment by ShadowJonathan** [comment]
User story: I have a workflow where I "peek" what a user has sent me, to quickly read through and get any important/critical points, and if its non-critical, I mark unread until the point that I haven't "processed" yet.

Basically, I make sure that I'll encounter it later, in exactly the same spot that I found it, so that next time, I can go through it with a better swing.

Marking the entire channel as unread does not allow me to mark a specific point of "read from here", and so there needs to be extra non-obvious context clues pointing to where I had already read, especially if its a quick discussion or a long back-n-forth DM that I needed to read in-depth at a later point.

**Comment by nunoperalta** [comment]
That's exactly my "story" as well. I check to see if there's anything urgent, or things I can reply quickly. Otherwise, mark as unread from the point I was before, so I can go back to it later.

(I use Element/Matrix with my small team, and not federated)

**Comment by colans** [comment]
I have the same story, but having the room unread is good enough for me because having something unprocessed in there is good enough for me.  I'll figure out what it is when I get back there.

> That's a screenshot of marking an entire room as unread, while this issue was about specific messages, no?

No.  See my update in the issue description (as the issue creator).

Let's do this:
1. Keep this open until you can mark the room unread in the mobile apps.
1. Create *another* follow-up issue for marking specific messages unread, which is blocked by [the ](https://github.com/matrix-org/matrix-spec/issues/1893).  (Can you even have blockers here?  I'm used to GitLab.)

I've updated the issue description with a list of remaining tasks.  Let's just work through it now.

**Comment by nunoperalta** [comment]
> Create another follow-up issue for marking specific messages unread

Ok, raised here, now: https://github.com/element-hq/element-meta/issues/2465

**Comment by TheJJ** [comment]
Element X on Android also supports the room unread-marking, so that's already good enough for now (I still have both apps, though).

**Comment by t3chguy** [comment]
Yeah the legacy apps won't be receiving new features such as this, only the X apps are in feature development

**Comment by colans** [comment]
Okay, so everything other than the mobile Apple stuff is checked off now.  Thanks for the updates!

**Comment by mcg-matrix** [comment]
> > Create another follow-up issue for marking specific messages unread
> 
> Ok, raised here, now: #2465

Well, that one's not about marking specific messages (= any specific message, individual messages) unread. Is that being tracked somewhere?

**Comment by nunoperalta** [comment]
> > > Create another follow-up issue for marking specific messages unread
> > 
> > 
> > Ok, raised here, now: #2465
> 
> Well, that one's not about marking specific messages (= any specific message, individual messages) unread. Is that being tracked somewhere?

I'm not sure it makes much sense to mark "specific/individual messages" as unread. I haven't seen that in any other app so far.

What I believe would make more sense is maybe some kind of "Private Pin" or "Star" functionality.
There is already "Pin", but I believe that pins to everyone, and not just for you.

**Comment by Jerome-Herbinet** [comment]
> > > > Create another follow-up issue for marking specific messages unread
> > > 
> > > 
> > > Ok, raised here, now: #2465
> > 
> > Well, that one's not about marking specific messages (= any specific message, individual messages) unread. Is that being tracked somewhere?
> 
> I'm not sure it makes much sense to mark "specific/individual messages" as unread. I haven't seen that in any other app so far.
> 
> What I believe would make more sense is maybe some kind of "Private Pin" or "Star" functionality.
> There is already "Pin", but I believe that pins to everyone, and not just for you.

@nunoperalta 
Unread message feature exists in Rocket Chat and it's very useful, everyday for my college and me.

**Comment by mcg-matrix** [comment]
> I'm not sure it makes much sense to mark "specific/individual messages" as unread. I haven't seen that in any other app so far.
> 
> What I believe would make more sense is maybe some kind of "Private Pin" or "Star" functionality. [...]

Love that idea! Let me find that old post... and shamelessly drop a self-reference: https://github.com/element-hq/element-meta/issues/347#issuecomment-1136132666 :-)

Is that already being tracked anywhere?
